server/src/postgres.rs

```diff
@@ -0,0 +1,464 @@
+use crate::{account::Account, database::SignalDatabase};
+use anyhow::{anyhow, bail, Result};
+use axum::{async_trait, extract};
+use common::{
+    signal_protobuf::Envelope,
+    web_api::{Device, DevicePreKeyBundle, UploadSignedPreKey},
+};
+use libsignal_core::{Aci, DeviceId, Pni, ProtocolAddress, ServiceId};
+use libsignal_protocol::{IdentityKey, PublicKey};
+use sqlx::{postgres::PgPoolOptions, Pool, Postgres};
+
+#[derive(Clone)]
+pub struct PostgresDatabase {
+    pool: Pool<Postgres>,
+}
+
+impl PostgresDatabase {
+    pub async fn connect() -> Result<Self> {
+        dotenv::dotenv().expect("Unable to load environment variables from .env file");
+        let db_url = std::env::var("DATABASE_URL").expect("Unable to read DATABASE_URL env var");
+        let pool = PgPoolOptions::new()
+            .max_connections(100)
+            .connect(&db_url)
+            .await
+            .map_err(|err| anyhow!(err))?;
+
+        Ok(Self { pool })
+    }
+}
+
+#[async_trait]
+impl SignalDatabase for PostgresDatabase {
+    async fn add_account(&self, account: Account) -> Result<()> {
+        sqlx::query!(
+            r#"
+            INSERT INTO
+                accounts (aci, pni, auth_token, identity_key)
+            VALUES
+                ($1, $2, $3, $4)
+            "#,
+            account.aci,
+            account.pni,
+            account.auth_token,
+            &*account.identity_key.serialize()
+        )
+        .execute(&self.pool)
+        .await
+        .map(|_| ())
+        .map_err(|err| err.into())
+    }
+
+    async fn get_account(&self, service_id: &ServiceId) -> Result<Account> {
+        let (aci, pni) = parse_service_id(service_id);
+
+        sqlx::query!(
+            r#"
+            SELECT 
+                aci, pni, auth_token, identity_key
+            FROM
+                accounts
+            WHERE
+                COALESCE(aci, '') = COALESCE($1, '')
+                OR COALESCE(pni, '') = COALESCE($2, '')
+            "#,
+            aci,
+            pni,
+        )
+        .fetch_one(&self.pool)
+        .await
+        .map(|row| Account {
+            aci: row.aci,
+            pni: row.pni,
+            auth_token: row.auth_token,
+            identity_key: IdentityKey::new(
+                PublicKey::deserialize(row.identity_key.as_slice()).unwrap(),
+            ),
+        })
+        .map_err(|err| err.into())
+    }
+
+    async fn update_account_aci(&self, service_id: &ServiceId, new_aci: Aci) -> Result<()> {
+        let (id_str, id) = pasrse_to_specific_service_id(service_id);
+
+        sqlx::query!(
+            r#"
+            UPDATE
+                accounts
+            SET
+                aci = $3
+            WHERE
+                $1 = $2
+            "#,
+            id_str,
+            id,
+            Some(new_aci.service_id_string())
```

Comment:
Please remove Some as it adds nothing
```suggestion
            new_aci.service_id_string()
```